src/internal/document_api/mod.rs

```diff
@@ -417,36 +418,46 @@ pub fn encrypt_document<'a, CR: rand::CryptoRng + rand::RngCore>(
             aes::encrypt_future(rng, &plaintext.to_vec(), *doc_sym_key.bytes()),
         )
         .and_then(move |(users, groups, encrypted_doc)| {
-            Ok({
-                let (group_errs, groups_with_key) = process_groups(groups);
-                let (user_errs, users_with_key) = process_users(users);
-                let self_grant = WithKey::new(
-                    UserOrGroup::User {
-                        id: auth.account_id.clone(),
-                    },
-                    user_master_pub_key.clone(),
-                );
-                let grant_list =
-                    [&[self_grant], &users_with_key[..], &groups_with_key[..]].concat();
-
-                // encrypt to all the users and groups
-                let (encrypt_errs, grants) = transform::encrypt_to_with_key(
-                    recrypt,
-                    &dek,
-                    &auth.signing_keys().into(),
-                    grant_list,
-                );
-
-                // squish all accumulated errors into one list
-                let other_errs = vec![
-                    group_errs,
-                    user_errs,
-                    encrypt_errs.into_iter().map(|e| e.into()).collect(),
-                ]
-                .into_iter()
-                .concat();
-                (grants, encrypted_doc, other_errs)
-            })
+            let (group_errs, groups_with_key) = process_groups(groups);
+            let (user_errs, users_with_key) = process_users(users);
+            let self_grant = WithKey::new(
+                UserOrGroup::User {
+                    id: auth.account_id.clone(),
+                },
+                user_master_pub_key.clone(),
+            );
+            let grant_list = if grant_to_author {
+                [&[self_grant], &users_with_key[..], &groups_with_key[..]].concat()
+            } else {
+                [&users_with_key[..], &groups_with_key[..]].concat()
+            };
+
+            if grant_list.is_empty() {
```

Comment:
I had to stare at this a while to make sense of it so it might be worth calling out that the reason you check this on the grant_list is because you would have already added self to the grant_list above if graunt_to_author was true.

Reply:
Maybe not a big deal, but it seems like maybe this validation ought to happen out at the call site in the `src/document.rs` file. We should be able to check that condition at that point right? I can't remember if we have any convention on this, but it seems like making sure that the arguments are valid before we call into the various `mod.rs` files might be good.